quickwit-cli/src/indexing/document_indexer.rs

```diff
@@ -0,0 +1,130 @@
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use crate::cli_command::IndexDataArgs;
+use tantivy::{schema::Schema, Document};
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::task;
+use tokio::task::JoinHandle;
+
+use crate::indexing::statistics::StatisticEvent;
+
+use crate::indexing::split::Split;
+
+pub async fn index_documents(
+    args: &IndexDataArgs,
+    mut document_receiver: Receiver<String>,
+    statistic_sender: Sender<StatisticEvent>,
+) -> anyhow::Result<Vec<Split>> {
+    //TODO get schema from docBuilder
+    let schema = Schema::builder().build();
+
+    let mut split_merge_tasks = vec![];
+
+    let mut current_split = Split::create(&args, schema.clone()).await?;
+    while let Some(raw_doc) = document_receiver.recv().await {
+        let doc_size = raw_doc.as_bytes().len();
+        let parse_result = parse_document(raw_doc);
+        let doc = match parse_result {
+            Ok(doc) => {
+                statistic_sender
+                    .send(StatisticEvent::NewDocument {
+                        size_in_bytes: doc_size,
+                        error: false,
+                    })
+                    .await?;
+                current_split.metadata.num_records += 1;
+                current_split.metadata.size_in_bytes += doc_size;
+                doc
+            }
+            Err(_) => {
+                statistic_sender
+                    .send(StatisticEvent::NewDocument {
+                        size_in_bytes: doc_size,
+                        error: true,
+                    })
+                    .await?;
+                current_split.metadata.num_parsing_errors += 1;
+                continue;
+            }
+        };
+
+        current_split.add_document(doc)?;
+        if current_split.has_enough_docs() {
+            let split = std::mem::replace(
+                &mut current_split,
+                Split::create(&args, schema.clone()).await?,
+            );
+
+            let finalise_split_task = finalize_split(split, statistic_sender.clone()).await?;
+            split_merge_tasks.push(finalise_split_task);
+        }
+    }
+
+    // Finalise last split if it has docs
+    if current_split.metadata.num_records > 0 {
+        let split = std::mem::replace(
+            &mut current_split,
+            Split::create(&args, schema.clone()).await?,
+        );
+
+        let finalise_split_task = finalize_split(split, statistic_sender.clone()).await?;
+        split_merge_tasks.push(finalise_split_task);
```

Comment:
Can you add a TODO.  we have no push back mechanism at this point...
Meaning if the finalize task cannot keep up with ingestion, we might end up with a lot of those running.